tests/unit/test_charm.py

```diff
@@ -181,6 +212,191 @@ def test_on_config_changed_iscsi(harness, mocker, iscsi_config):
+
+    assert harness.charm._get_initiator_name_from_file() == "new-iqn"
+
+    # assert that a random iqn isn't generated
+    assert call("/sbin/iscsi-iname") not in mock_getoutput.mock_calls
+    # check if file is being rendered with expected content
+    mock_write_text.assert_has_calls(
+        [
+            call(expected_initiator_content),
+        ]
+    )
+    assert isinstance(harness.charm.unit.status, ActiveStatus)
+
+
+def test_iscsi_with_initiatorname_not_rendered(mocker, harness, iscsi_config):
+    """Test iscsi configuration when initiatorname.iscsi is present and not rendered again.
+
+    Tests 2 scenarios when the file isn't rendered:
+    1. The initiator name from initiatorname.iscsi contains the same value as the one in the
+    initiator-dictionary.
+    2. The initiator-dictionary doesn't have an initiator name and one is present in the
+    initiatorname.iscsi file.
+
+    """
+    mocker.patch("charm.utils.is_container", return_value=False)
+    mock_getoutput = mocker.patch(
+        "charm.subprocess.getoutput", return_value="iqn.2020-07.canonical.com:lun1"
+    )
+    mocker.patch("charm.socket.getfqdn", return_value="testhost.testdomain")
+    mocker.patch("charm.subprocess.check_call")
+    mocker.patch("charm.subprocess.check_output")
+    mocker.patch("charm.StorageConnectorCharm._configure_deferred_restarts")
+    mocker.patch("charm.StorageConnectorCharm._defer_service_restart")
+    mocker.patch("charm.Path.chmod")
+
+    # test scenario 1
+    # initiator-dictionary configured with same name as the one present in initiatorname.iscsi
+    expected_initiator_content = dedent(
+        """\
+        ###############################################################################
+        # [ WARNING ]
+        # configuration file maintained by Juju
+        # local changes will be overwritten.
+        #
+        # DO NOT EDIT OR REMOVE THIS FILE!
+        # If you remove this file, the iSCSI daemon will not start.
+        # If you change the InitiatorName, existing access control lists
+        # may reject this initiator.  The InitiatorName must be unique
+        # for each iSCSI initiator.  Do NOT duplicate iSCSI InitiatorNames.
+        ###############################################################################
+        InitiatorName=iqn.2020-07.canonical.com:lun1"""
+    )
+    initiator_content_with_same_name = expected_initiator_content
+    harness.charm.ISCSI_INITIATOR_NAME.write_text(initiator_content_with_same_name)
+    mock_write_text = mocker.patch("charm.Path.write_text")
+
+    harness.charm._stored.installed = True
+    harness.update_config(iscsi_config)
+
+    assert harness.charm._get_initiator_name_from_file() == "iqn.2020-07.canonical.com:lun1"
+    # random iqn not generated
+    assert call("/sbin/iscsi-iname") not in mock_getoutput.mock_calls
+    # initiatorname.iscsi not rendered again
+    assert call(expected_initiator_content) not in mock_write_text.mock_calls
```

Comment:
I will check the modified time of the file is possible. 
or
If mock_write_text was executed, then I will say it is failed.

Reply:
> If mock_write_text was executed, then I will say it is failed.

Isn't this what the test is currently doing? If `mock_write_text` had a call with the `expected_initiator_content` then we know that the file is being rendered again and the assertion would fail.

Reply:
Yes, but I prefer to validate if `mock_write_text` had a call is good enough. We don't need to validate the full content.

Reply:
The problem with that is when the config changed hook is executed, there will be a number of other calls to `Path.write_text` and subsequently for `mock_write_text`. So it won't be enough to just check whether `mock_write_text` had been called.

Maybe I can do something similar to what I do later in the same test, ie, assert that the number of calls to `mock_write_text` is  2 (if it was called by function being test, it would've been 3).